nova/api/openstack/placement/lib.py

```diff
@@ -31,12 +32,15 @@ def __init__(self, use_same_provider=True, resources=None,
         :param forbidden_traits: A set of { trait_name, ... }
         :param member_of: A list of [ [aggregate_UUID],
                                       [aggregate_UUID, aggregate_UUID] ... ]
+        :param ignore_consumers: A list of UUIDs of consumers that should be
+                ignored when finding allocation candidates.
         """
         self.use_same_provider = use_same_provider
         self.resources = resources or {}
         self.required_traits = required_traits or set()
         self.forbidden_traits = forbidden_traits or set()
         self.member_of = member_of or []
+        self.ignore_consumers = ignore_consumers or []
```

Comment:
Could this be a `set()`?

Two counter-arguments that I see:

1. It's too much work because it's unlikely that there's duplication in the list. (But it's possible.)
1. Does SQLAlchemy's `in_()` play well with sets? I tried answering this briefly but couldn't quickly enough construct a query in IPython before I got annoyed with sqlalchemy, again...

```suggestion
        self.ignore_consumers = ignore_consumers or set()
```

Reply:
How does that make sense? If we get a value, it's coming in as a list - as documented. If we don't get a value (i.e. `None`), we change the datatype used to a set instead of a list?

Yeah, things can be in there double, but it shouldn't matter for our query if we exclude it once or twice.

Reply:
Yeah, sorry, I meant making it a set not just in the default here, but elsewhere as well.

And it doesn't have a big impact whether it's a list or a set, just that a set type documents better what it is, imo. But a list is fine as well.

Reply:
then I guess it should be `= set(ignore_consumers) or set()`? So that we handle it as a set internally. Or do you suggest we should reject duplicates at the api level?

Reply:
Yes something like that.

I just raised this to get some input from you guys, if this is even a concern. Because I think the code behaves the same with a list. It's purely to reveal the intent of the code, not a fix for anything.

Reply:
If we want it to be a set, we want it to be a set everywhere. This means, we would have to change https://github.com/sapcc/nova/pull/176/files#diff-57bafb2379c07e0d5c0175b2e40c0bc7b9b6aa78029ce88bba10c7015a4d8bdfR504 to be a set, because that's where we retrieve the data from the request and validate it. Afterwards, it should be in the form we want it - being it a set or a list.

Reply:
Question remains: Is it worth the effort?

Reply:
I don't think so.